src/main/java/com/iota/iri/cache/impl/CacheImpl.java

```diff
@@ -0,0 +1,172 @@
+package com.iota.iri.cache.impl;
+
+import com.iota.iri.cache.Cache;
+import com.iota.iri.cache.CacheConfiguration;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentMap;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.MapMaker;
+
+/**
+ * Cache operations
+ */
+public class CacheImpl<K, V> implements Cache<K, V> {
+
+    // Config
+    private final CacheConfiguration cacheConfiguration;
+
+    // Actual store
+    private final ConcurrentMap<K, V> strongStore;
+    //weak store. Eligible for GC
+    private final ConcurrentMap<K,V> weakStore;
+    private final ConcurrentLinkedQueue<K> evictionQueue;
+
+    // stats
+    private int cacheHits = 0;
+    private int cacheMisses = 0;
+
+    /**
+     * Constructor
+     * 
+     * @param cacheConfiguration The configuration to use to instantiate this cache
+     */
+    public CacheImpl(CacheConfiguration cacheConfiguration) {
+        this.cacheConfiguration = cacheConfiguration;
+        this.strongStore = new MapMaker().concurrencyLevel(5).makeMap();
+        this.weakStore = new MapMaker().concurrencyLevel(5).weakValues().makeMap();
+        this.evictionQueue = new ConcurrentLinkedQueue<>();
+    }
+
+    @Override
+    public V get(K key) {
+        if(key == null){
+            return null;
+        }
+        V value = strongStore.get(key);
+        if(value == null){
+            //convert back to strong and remove from weak map
+            if(weakStore.containsKey(key)){
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [These nested if statements could be combined](https://app.codacy.com/app/GalRogozinski/iri/pullRequest?prid=4567264)